Behavior policies added to the Policy Cookbook. (#25)
hivemq/hivemq-policy-cookbooks: behavior-validation/publish-duplicate/README.md

# Disconnect a client sending duplicate messages

> This cookbook is about to disconnect a client when it sends a message similar to the previous one.

Reviewer: ```suggestion

This cookbook is about disconnecting a client when it sends a PUBLISH message identical to the previous one it sent.
```

### Use-case
> As a developer, I want to disconnect all clients that send duplicate messages.

For this use-case, a behavior policy is required.

### Policy

> Consider the following behavior policy that disconnects a client if the received published message is similar to the previous one.

Reviewer: ```suggestion
Consider the following behavior policy that disconnects a client if the received published message has an identical payload and topic to the previous one.
```

> This policy uses `System.log` function to log explanatory messages on each state transition and `Mqtt.disconnect` function to disconnect a client if message count goest above 10.

Reviewer: This appears to be the description for the other cookbook.

`policy.json`:
```json
{
  "id": "duplicate-policeman",
  "matching": {

>     "clientIdRegex": ".*"

Reviewer: Both examples use `.*`. I think it would be nice to have one use a more specific regex to demonstrate how you can match only certain clients + a brief explanation about that in the cookbook.

  },
  "behavior": {
    "id": "Publish.duplicate",
    "arguments": {}
  },
  "onTransitions": [
    {
      "fromState": "Initial",
      "toState": "Connected",
      "Mqtt.OnInboundConnect": {
        "pipeline": [
          {
            "id": "logFunctionTransitionDebug",
            "functionId": "System.log",
            "arguments": {

>               "level": "INFO",

Reviewer: Should this be at the DEBUG level?

              "message": "Behavior policy ${policyId}: ${clientId} transitioned from ${fromState} to ${toState} on ${triggerEvent} at ${timestamp}"
            }
          },
          {
            "id": "logFunctionExplanationMessage",
            "functionId": "System.log",
            "arguments": {
              "level": "INFO",
              "message": "Ok, we've got a client connected. Waiting for a first PUBLISH message to start detecting duplicates."
            }
          }
        ]
      }
    },
    {
      "fromState": "Connected",
      "toState": "NotDuplicated",
      "Mqtt.OnInboundPublish": {
        "pipeline": [
          {
            "id": "logFunctionTransitionDebug",
            "functionId": "System.log",
            "arguments": {

>               "level": "INFO",

Reviewer: DEBUG level

              "message": "Behavior policy ${policyId}: ${clientId} transitioned from ${fromState} to ${toState} on ${triggerEvent} at ${timestamp}"
            }
          },
          {
            "id": "logFunctionExplanationMessage",
            "functionId": "System.log",
            "arguments": {
              "level": "INFO",
              "message": "And here is the first PUBLISH message. From now on we are detecting duplicates."
            }
          }
        ]
      }
    },
    {
      "fromState": "NotDuplicated",
      "toState": "Duplicated",
      "Mqtt.OnInboundPublish": {
        "pipeline": [
          {
            "id": "logFunctionTransitionDebug",
            "functionId": "System.log",
            "arguments": {

>               "level": "INFO",

Reviewer: DEBUG level

              "message": "Behavior policy ${policyId}: ${clientId} transitioned from ${fromState} to ${toState} on ${triggerEvent} at ${timestamp}"
            }
          },
          {
            "id": "logFunctionExplanationMessage",
            "functionId": "System.log",
            "arguments": {
              "level": "WARN",
              "message": "Duplicate message detected! ${clientId} will be disconnected!"
            }
          },
          {
            "id": "disconnectFunction",
            "functionId": "Mqtt.disconnect",
            "arguments": {}
          }
        ]
      }
    }
  ]
}
```

To upload `policy.json` to the broker, run the following command:

```bash
mqtt hivemq behavior-policy create --file policy.json
```

The policy is now applied and all incoming MQTT messages are subject to validation.

To delete the policy, run the following command:

```bash
mqtt hivemq behavior-policy delete --id duplicate-policeman
```

### Policy output

> Excessive amount of `System.log` calls used in the policy produces an output that facilitates understanding of the policy's behavior.

Reviewer: "excessive" 😆 

```suggestion
The abundant use of `System.log` in the policy produces a log output that facilitates understanding of the policy's behavior.
```

### Policy

Consider the following behavior policy that disconnects a client if the received published message has an identical payload and topic to the previous one.

> This policy uses `System.log` function to log explanatory messages on each state transition and `Mqtt.disconnect` function to disconnect a client if received message is duplicate.

Reviewer: ```suggestion
This policy uses `System.log` function to log explanatory messages on each state transition and `Mqtt.disconnect` function to disconnect a client if the received message is a duplicate.
```

> Spot the `clientIdRegex` that makes this policy apply only to the clientIds starting with `hmq` with the regex `^hmq.*`.

Reviewer: ```suggestion
Note the `clientIdRegex` that causes this policy to apply only to the clientIds starting with `hmq` with the regex `^hmq.*`.
```
